### PR TITLE
Make aircraft occupy cells when landed.

### LINKED ISSUE #2018: more then 1 unit at FIX
Sent HARV to FIX (Service Depot), then sent YAK and it landed on HARV. Looks like both repaired, then HARV left FIX and another  HARV used depot without problems while YAK was in there.

Screenshot:  http://files.ihptru.net/misc/Screenshot-18_1.png

Replay: Not required - Reproducable

To reproduce: happens with any combination of AIR / NON AIR units

Version: 2012-0315
## <bountysource-plugin>

Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/98329-more-then-1-unit-at-fix?utm_campaign=plugin&utm_content=tracker%2F36085&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F36085&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### LINKED ISSUE #4605: Chinook doesn't block vehicles when landed
Affects RA. May or may not affect TD.
## <bountysource-plugin>

Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/1401344-chinook-doesn-t-block-vehicles-when-landed?utm_campaign=plugin&utm_content=tracker%2F36085&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F36085&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### LINKED ISSUE #14114: Two yaks landing on one air field
[OpenRA-2017-10-04T192055Z.orarep.zip](https://github.com/OpenRA/OpenRA/files/1357169/OpenRA-2017-10-04T192055Z.orarep.zip)

While I was trying to find out the cause of #14102 I noticed that two yaks were landing and reloading on the same air field the same time.

It happens in the replay at ~11:45.
release-20170527

### LINKED ISSUE #14867: 2 aircraft landing on 1 airfield
### Issue Summary
In the replay at 11 minutes 2 aircraft land on 1 airfield after a new airfield was built. Despite the fact one of the aircraft was already armed, its my understanding this isnt intended.

#### Additional Information:
I am unsure of how to reproduce this bug. I attempted myself to show in an isolated replay but I don't know what caused it. It could be possibly a combination of having given a command to reload, placing a new airfield and giving the command to reload again.

- OpenRA Replays
[2 planes 1 airfield 11minutes.zip](https://github.com/OpenRA/OpenRA/files/1771777/2.planes.1.airfield.11minutes.zip)

It occurs at 11 minutes in my base (pink)

### LINKED ISSUE #15131: Removing Targetable@AIRBORNE: from plane aircraft will make them all able to land in a single helipad.
Havent tested with helicopters, not sure if needed.
aurora

### PR DESCRIPTION
**Pull request by tovl**
This makes aircraft occupy space on the `ActorMap` when landed. This prevents other ground units from moving into the same space.

Fixes #4605.
Fixes #2018.
Fixes #15131
Fixes #14114
Fixes #14867

This makes fixed-wing aircraft check for blocking actors as well. If they find the repair depot (or any other landing spot) blocked, they will circle and wait for it to be unblocked. Aircraft will try to reserve the required cells when landing is initiated.

### CODE REVIEW

**Comment by pchote** [comment]
Can you comment on how this interacts with crates and mines?

File: mods/cnc/rules/aircraft.yaml [hunk 1]
```diff
@@ -68,6 +68,7 @@ HELI:
 	Aircraft:
 		TurnSpeed: 7
 		Speed: 180
+		LandableTerrainTypes: Clear,Rough,Road,Ore,Beach,Tiberium,BlueTiberium
```
Comment (teinarss) [inline mods/cnc/rules/aircraft.yaml]: Should prolly be added to the transport heli too, Chinook.

Comment (tovl) [reply mods/cnc/rules/aircraft.yaml]: This was copy-pasted from the transport heli; See Line 18.

**Comment by tovl** [comment]
> Can you comment on how this interacts with crates and mines?

Helicopters cannot land on mines or crates. This is true in bleed as well.

**Comment by pchote** [comment]
Is there any prospect for fixing that here? Aircraft not interacting with the ActorMap was always the main blocker for supporting those.

**Comment by tovl** [comment]
I think that should be handled in a follow-up PR.

File: mods/cnc/rules/aircraft.yaml (hunk 1 above)
Comment (reaperrr) [inline mods/cnc/rules/aircraft.yaml]: All aircraft in TD except transport helis have `LandWhenIdle: false` and can only land on helipads, and the terrain type shouldn't matter there, so these yaml changes should be unnecessary.

File: mods/ra/rules/aircraft.yaml [hunk 2]
```diff
@@ -118,6 +118,7 @@ MIG:
 		Speed: 223
 		RepulsionSpeed: 40
 		MaximumPitch: 56
+		LandableTerrainTypes: Clear,Rough,Road,Beach
```
Comment (reaperrr) [inline mods/ra/rules/aircraft.yaml]: All aircraft in RA except transport helis have `LandWhenIdle: false` and can only land on airfields/helipads or repair depots, and the terrain type shouldn't matter there, so these yaml changes should be unnecessary.

File: mods/cnc/rules/aircraft.yaml (hunk 1 above)
Comment (tovl) [reply mods/cnc/rules/aircraft.yaml]: Unfortunately, that is not true. In order to fix #2018 aircraft will have to explicitly check whether they can land on a given cell even when called from ReturnToBase. when calling CanLand they will then also check the underlying terrain. I admit this is a bit hacky. There is probably a better solution to this.

Even so, I think helicopters should at least in theory be able to land outside of the helipad even if currently there is no way to order them to do so. They could in the original and conceptually it makes sense.

**Comment by tovl** [comment]
So, I've decided to add the crushing logic to aircraft anyway. This will allow transport helicopters to interact with crates and mines. It will also allow them to squash infantry unlucky enough to stand right under them when landing.

File: mods/cnc/rules/aircraft.yaml (hunk 1 above)
Comment (tovl) [reply mods/cnc/rules/aircraft.yaml]: Ok. I've added an extra check in `CanLand` so this hack is not necessary anymore. Removed `LandableTerrainTypes` from non-vtol craft. I've left it in for helicopters. I guess that's just a matter of taste. I will remove it if you find it really objectionable.

File: mods/ra/rules/aircraft.yaml (hunk 2 above)
Comment (tovl) [reply mods/ra/rules/aircraft.yaml]: Ok. I've added an extra check in `CanLand` so this hack is not necessary anymore. Removed `LandableTerrainTypes` from non-vtol craft. I've left it in for helicopters. I guess that's just a matter of taste. I will remove it if you find it really objectionable.

**Comment by matjaeck** [comment]
Is this intended to fix helicopters, too? For example 2 helis on one pad or vehicle and heli on fix.

**Comment by tovl** [comment]
It should work for all types of aircraft.

**Review by ghost — COMMENTED** [review]
Crate collection with Chinooks works well (but not all crate effects seem to work for aircraft - iron curtain crate did not have any effect). Crushing infantry with Chinooks is fun, too. For planes everything seems to work like intended.

Findings from a quick test:
<details>
  <summary>helicopters and vehicles can still repair simultaneously on service depots</summary>

helitankfix

</details>

<details>
  <summary>multiple helicopters at one pad</summary>

 If you send 2 helicopters to rearm at a helipad, and then hit deploy when the first landed, the second heli will land on top of it. Rearming of the first heli will be stopped and it will only take off and land again when the 2nd heli has finished rearming.

2xhelipad

</details>

It can happen that 2 helicopters try to land at the same spot at the same time and then land on top of the other one. It's an edge case though that is unlikely to happen in normal games I think.

**Comment by tovl** [comment]
Helicopters should be fixed now.

**Comment by matjaeck** [comment]
Great, I'll also test it against these issues, but not tonight: #14114, #14867, #15131

**Comment by tovl** [comment]
Update:
- Fixed chinook being unable to unload cargo due to occupying space.
- Added missing `CrushDamageTypes` yaml parameter.
- aircraft can now nudge friendly units out of the way when landing.

**Comment by tovl** [comment]
Rebased.

**Review by ghost — COMMENTED** [review]
#15131 looks fixed, haven't watched the replays of  #14114 and #14867 but couldn't reproduce the issue when testing in RA.

**Review by ghost — DISMISSED** [review]

**Comment by tovl** [comment]
Rebased.

**Comment by matjaeck** [comment]
:+1:

**Comment by reaperrr** [comment]
Transport heli constantly lands and takes off again.

**Comment by matjaeck** [comment]
> Transport heli constantly lands and takes off again.

I assumed this is the already fixed #16368.

**Comment by reaperrr** [comment]
I rebased onto latest bleed, but I still get that bug, so this PR appears to regress it again.

**Comment by tovl** [comment]
Fixed.

File: mods/cnc/rules/aircraft.yaml (hunk 1 above)
Comment (reaperrr) [reply mods/cnc/rules/aircraft.yaml]: Well, 'objectionable' sounds a bit too harsh, we just try to avoid keeping lines around that won't have any effect on that actor.
That being said, as a compromise you could move this to the `^Helicopter` default instead.

File: OpenRA.Mods.Common/Traits/Air/Aircraft.cs [hunk 3]
```diff
@@ -640,6 +681,35 @@ public void SetPosition(Actor self, WPos pos)
 				OnCruisingAltitudeReached();
 			else if (!isCruising && cruising)
 				OnCruisingAltitudeLeft();
+			FinishedMoving(self);
```
Comment (reaperrr) [inline OpenRA.Mods.Common/Traits/Air/Aircraft.cs]: I'd prefer an empty line above this for better readability.

File: mods/ra/rules/aircraft.yaml (hunk 2 above)
Comment (reaperrr) [reply mods/ra/rules/aircraft.yaml]: See my comment on cnc above.

**Review by reaperrr — DISMISSED** [review]
Looks good to me apart from minor nits

**Comment by reaperrr** [comment]
Needs a rebase, too.

File: OpenRA.Mods.Common/Traits/Air/Aircraft.cs (hunk 3 above)
Comment (tovl) [reply OpenRA.Mods.Common/Traits/Air/Aircraft.cs]: Fixed.

**Comment by tovl** [comment]
Rebased.

**Comment by reaperrr** [comment]
@tovl I guess you overlooked it (understandable since it's quite a bit up in the comments):

Regarding the `LandableTerrain` of RA/TD helicopters (and nearly all TS aircraft):
> Well, 'objectionable' sounds a bit too harsh, we just try to avoid keeping lines around that won't have any effect on that actor.
That being said, as a compromise you could move this to the `^Helicopter` default instead.

**Comment by tovl** [comment]
Yes, I missed that. I've removed these lines for now, because as you said they do not have any effect currently.

**Comment by reaperrr** [comment]
```
[18:44] <+reaperrr> pchote: what was the technical reason again for preferably not using writeable public booleans, ints and so on?
[18:49] <+pchote> in what context?
[18:52] <+reaperrr> https://github.com/OpenRA/OpenRA/pull/16315/commits/30e77af6049e92d34adf22f7c4ed7a5e2ac47c57#diff-e8b231bb90082b00e18a95d9ae6a7ef4R187
[18:54] <+pchote> basically want to expose as small a public surface area as possible
[18:56] <+pchote> that LandingInitiated doesn't seem to be used at all?
```

**Comment by pchote** [comment]
`LandingCell` could perhaps be kept private by passing the cell as an argument to `AddInfluence`.

**Comment by tovl** [comment]
Fixed.

**Review by reaperrr — APPROVED** [review]

**Comment by reaperrr** [comment]
In addition to #16417, the changes here - in particular to Land/HeliLand - are also a dependency for updating #16365, as well as a few of my local PRs, so it would be good to get this merged soon (as long as there are no remaining issues, of course).

File: OpenRA.Mods.Common/Traits/Air/Aircraft.cs [hunk 4]
```diff
@@ -528,6 +548,41 @@ public bool CanRepairAt(Actor host)
 			return repairable != null && repairable.Info.RepairActors.Contains(host.Info.Name) && self.GetDamageState() != DamageState.Undamaged;
 		}
 
+		bool IsBlockedBy(Actor self, Actor otherActor, Actor ignoreActor)
+		{
+			// We are not blocked by the actor we are ignoring.
+			if (otherActor == ignoreActor)
```
Comment (reaperrr) [inline OpenRA.Mods.Common/Traits/Air/Aircraft.cs]: If you change this to `if (otherActor == self || otherActor == ignoreActor)` you can drop the work-arounds...

File: OpenRA.Mods.Common/Traits/Air/Aircraft.cs [hunk 5]
```diff
@@ -575,7 +630,7 @@ protected virtual void OnBecomingIdle(Actor self)
 			}
 			else if (!Info.CanHover && !atLandAltitude)
 				self.QueueActivity(new FlyCircle(self, -1, Info.IdleTurnSpeed > -1 ? Info.IdleTurnSpeed : TurnSpeed));
-			else if (atLandAltitude && !CanLand(self.Location) && ReservedActor == null)
+			else if (atLandAltitude && !CanLand(self.Location, self) && ReservedActor == null)
```
Comment (reaperrr) [inline OpenRA.Mods.Common/Traits/Air/Aircraft.cs]: ...here...

File: OpenRA.Mods.Common/Traits/Cargo.cs [hunk 6]
```diff
@@ -213,7 +213,7 @@ public bool CanUnload(bool immediate = false)
 					return false;
 			}
 
-			return !IsEmpty(self) && (aircraft == null || aircraft.CanLand(self.Location))
+			return !IsEmpty(self) && (aircraft == null || aircraft.CanLand(self.Location, self))
```
Comment (reaperrr) [inline OpenRA.Mods.Common/Traits/Cargo.cs]: ...and here.

**Review by reaperrr — CHANGES_REQUESTED** [review]
Thinking about it, the aircraft should *never* block itself from landing.

File: OpenRA.Mods.Common/Activities/Air/Fly.cs [hunk 7]
```diff
@@ -105,6 +105,9 @@ public override Activity Tick(Actor self)
 				soundPlayed = true;
 			}
 
+			if (self.IsAtGroundLevel())
```
Comment (pchote) [inline OpenRA.Mods.Common/Activities/Air/Fly.cs]: Can you please add a comment here like `// We are taking off, so remove ground influence` to help reduce confusion?

File: OpenRA.Mods.Common/Activities/Air/HeliFly.cs [hunk 8]
```diff
@@ -98,6 +98,9 @@ public override Activity Tick(Actor self)
 				soundPlayed = true;
 			}
 
+			if (self.IsAtGroundLevel())
```
Comment (pchote) [inline OpenRA.Mods.Common/Activities/Air/HeliFly.cs]: Likewise here.

File: OpenRA.Mods.Common/Activities/Air/HeliLand.cs [hunk 9]
```diff
@@ -19,26 +19,42 @@ public class HeliLand : Activity
...
+			}
 
-			if (requireSpace && !aircraft.CanLand(self.Location))
-				return this;
+			if (requireSpace && !landingInitiated)
+			{
+				var landingCell = self.Location;
+				if (!aircraft.CanLand(landingCell, ignoreActor))
+				{
+					self.NotifyBlocker(landingCell);
```
Comment (pchote) [inline OpenRA.Mods.Common/Activities/Air/HeliLand.cs]: Pinging the blocker every tick isn't great for performance. Please add a delay between retries like you do in `Land`.

File: OpenRA.Mods.Common/Traits/Air/Aircraft.cs [hunk 10]
```diff
@@ -493,7 +498,13 @@ public int MovementSpeed
 			get { return Util.ApplyPercentageModifiers(Info.Speed, speedModifiers); }
 		}
 
-		public Pair<CPos, SubCell>[] OccupiedCells() { return NoCells; }
+		public Pair<CPos, SubCell>[] OccupiedCells()
+		{
+			if (!self.IsAtGroundLevel())
```
Comment (pchote) [inline OpenRA.Mods.Common/Traits/Air/Aircraft.cs]: `OccupyCells` can be called by arbitrary things at arbitrary times, so we should still be returning `NoCells` during normal flight to avoid bogosity.

File: OpenRA.Mods.Common/Traits/Air/Aircraft.cs [hunk 11]
```diff
@@ -506,13 +517,22 @@ public WVec FlyStep(int speed, int facing)
...
-		public bool CanLand(CPos cell)
+		public bool CanLand(CPos cell, Actor ignoreActor = null)
 		{
 			if (!self.World.Map.Contains(cell))
 				return false;
 
-			if (self.World.ActorMap.AnyActorsAt(cell))
-				return false;
+			foreach (var otherActor in self.World.ActorMap.GetActorsAt(cell))
+			{
```
Comment (pchote) [inline OpenRA.Mods.Common/Traits/Air/Aircraft.cs]: Style nit: these braces can be omitted

File: OpenRA.Mods.Common/Traits/Air/Aircraft.cs [hunk 12]
```diff
@@ -506,13 +517,22 @@ public WVec FlyStep(int speed, int facing)
...
+				if (IsBlockedBy(self, otherActor, ignoreActor))
+					return false;
+			}
+
+			foreach (var otherActor in self.World.ActorMap.GetActorsAt(cell))
+			{
```
Comment (pchote) [inline OpenRA.Mods.Common/Traits/Air/Aircraft.cs]: These ones too.

File: OpenRA.Mods.Common/Traits/Air/Aircraft.cs [hunk 13]
```diff
@@ -528,6 +548,41 @@ public bool CanRepairAt(Actor host)
...
+		bool IsBlockedBy(Actor self, Actor otherActor, Actor ignoreActor)
+		{
+			// We are not blocked by the actor we are ignoring.
+			if (otherActor == ignoreActor)
+				return false;
+
+			// If self is null, we don't have a real actor - we're just checking what would happen theoretically.
+			// In such a scenario - we'll just assume any other actor in the cell will block us by default.
+			// If we have a real actor, we can then perform the extra checks that allow us to avoid being blocked.
+			if (self == null)
```
Comment (pchote) [inline OpenRA.Mods.Common/Traits/Air/Aircraft.cs]: This is inside `Aircraft` (not `AircraftInfo`), so `self` is guaranteed by convention to be non-null.

File: OpenRA.Mods.Common/Traits/Air/Aircraft.cs [hunk 14]
```diff
@@ -615,16 +670,61 @@ public void SetPosition(Actor self, WPos pos)
...
+			if (!self.IsAtGroundLevel())
+				return;
+
+			var actors = self.World.ActorMap.GetActorsAt(TopLeft).Where(a => a != self).ToList();
+			if (!AnyCrushables(actors))
+				return;
+
+			var notifiers = actors.SelectMany(a => a.TraitsImplementing<INotifyCrushed>().Select(t => new TraitPair<INotifyCrushed>(a, t)));
+			foreach (var notifyCrushed in notifiers)
+				notifyCrushed.Trait.OnCrush(notifyCrushed.Actor, self, Info.Crushes);
```
Comment (pchote) [inline OpenRA.Mods.Common/Traits/Air/Aircraft.cs]: We need to call `WarnCrush` first, probably when starting to land to match `Mobile.EnteringCell`, so that enemy units have a chance to run away.

**Review by pchote — COMMENTED** [review]
Looking good, but a few comments:

File: OpenRA.Mods.Common/Traits/Air/Aircraft.cs (hunk 4 above)
Comment (tovl) [reply OpenRA.Mods.Common/Traits/Air/Aircraft.cs]: Yep, I already did this is #16417. But it makes more sense to include it here. Changed.

File: OpenRA.Mods.Common/Traits/Air/Aircraft.cs (hunk 5 above)
Comment (tovl) [reply OpenRA.Mods.Common/Traits/Air/Aircraft.cs]: Fixed.

File: OpenRA.Mods.Common/Traits/Cargo.cs (hunk 6 above)
Comment (tovl) [reply OpenRA.Mods.Common/Traits/Cargo.cs]: Fixed.

File: OpenRA.Mods.Common/Activities/Air/Fly.cs (hunk 7 above)
Comment (tovl) [reply OpenRA.Mods.Common/Activities/Air/Fly.cs]: Done.

File: OpenRA.Mods.Common/Activities/Air/HeliFly.cs (hunk 8 above)
Comment (tovl) [reply OpenRA.Mods.Common/Activities/Air/HeliFly.cs]: Done.

File: OpenRA.Mods.Common/Activities/Air/HeliLand.cs (hunk 9 above)
Comment (tovl) [reply OpenRA.Mods.Common/Activities/Air/HeliLand.cs]: Done.

File: OpenRA.Mods.Common/Traits/Air/Aircraft.cs (hunk 10 above)
Comment (tovl) [reply OpenRA.Mods.Common/Traits/Air/Aircraft.cs]: Okay. Tied this to `RemoveInfluence` to mirror `AddInfluence`.

File: OpenRA.Mods.Common/Traits/Air/Aircraft.cs (hunk 11 above)
Comment (tovl) [reply OpenRA.Mods.Common/Traits/Air/Aircraft.cs]: Fixed.

File: OpenRA.Mods.Common/Traits/Air/Aircraft.cs (hunk 12 above)
Comment (tovl) [reply OpenRA.Mods.Common/Traits/Air/Aircraft.cs]: Fixed.

File: OpenRA.Mods.Common/Traits/Air/Aircraft.cs (hunk 13 above)
Comment (tovl) [reply OpenRA.Mods.Common/Traits/Air/Aircraft.cs]: Oops. That's what I get for shamelessly copy-pasting. Fixed.

File: OpenRA.Mods.Common/Traits/Air/Aircraft.cs (hunk 14 above)
Comment (tovl) [reply OpenRA.Mods.Common/Traits/Air/Aircraft.cs]: Added `EnteringCell` method to `Aircraft` and called upon landing initiation.

Comment (matjaeck) [reply OpenRA.Mods.Common/Traits/Air/Aircraft.cs]: Something is not working, they always die.

Comment (tovl) [reply OpenRA.Mods.Common/Traits/Air/Aircraft.cs]: Should be fixed now.

**Review by ghost — APPROVED** [review]
LGTM

**Review by reaperrr — APPROVED** [review]

**Comment by reaperrr** [comment]
https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/9e39b86618fd1fa5c51285350436785cbe0d2ce1...324d6479288abbd7d7b659abd2c6f4a65dca347c
